### PR TITLE
[JENKINS-59793] Avoid hanging jobs with faulty SubTasks

### PR DESCRIPTION
**Pull request by MRamonLeon**
See [JENKINS-59793](https://issues.jenkins-ci.org/browse/JENKINS-59793).

I have created a test to reproduce the status reached by several private Jenkins instances which have thousand of *Metric Plugin* threads (_QueueSubTaskMetrics_) running. It's all because if a Job has faulty SubTasks (throwing an unmanaged exception), the `future` object is not set in the _**WorkUnitContext#synchronizeEnd**_ method. It may also happen in WorkUnitContext#synchronizeStart but I have not reproduced this case).

Whatever code that keeps waiting for the start or the end of the build will stay there forever. The Metrics Plugin is impacted by this here:
* Waiting for the build to finish: https://github.com/jenkinsci/metrics-plugin/blob/24bf92ebd59095d8f77b2552696b3f210a024bdc/src/main/java/jenkins/metrics/impl/JenkinsMetricProviderImpl.java#L1142

Called from: https://github.com/jenkinsci/metrics-plugin/blob/24bf92ebd59095d8f77b2552696b3f210a024bdc/src/main/java/jenkins/metrics/impl/JenkinsMetricProviderImpl.java#L944

I will let the CI fail and then I will push the fix.

### Proposed changelog entries

* Entry 1: Improve the resilience to faulty subtask contributors which may keep the build running forever.

### Submitter checklist

- [X] JIRA issue is well described
- [X] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [X] Appropriate autotests or explanation to why this change has no tests
- [N/A] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@daniel-beck @varyvol @batmat @alecharp @rsandell @fcojfernandez @stephenc

### CODE REVIEW

**Comment by MRamonLeon** [comment]
Screenshot from 2019-11-07 20-50-24

```
Stacktrace
java.util.concurrent.TimeoutException
	at hudson.remoting.AsyncFutureImpl.get(AsyncFutureImpl.java:97)
	at hudson.model.queue.BuildKeepsRunningWhenFaultySubTasksTest.buildDoesntFinishWhenSubTaskFails(BuildKeepsRunningWhenFaultySubTasksTest.java:46)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:600)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:834)
...
  10.292 [id=32]	SEVERE	hudson.model.Executor#finish1: Executor threw an exception
java.lang.ArrayIndexOutOfBoundsException: My unexpected exception
	at hudson.model.queue.BuildKeepsRunningWhenFaultySubTasksTest$FailingSubTaskContributor$1$1.run(BuildKeepsRunningWhenFaultySubTasksTest.java:65)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:427)
```

**Review by res0nance — APPROVED** [review]
LGTM :+1:

File: test/src/test/java/hudson/model/queue/BuildKeepsRunningWhenFaultySubTasksTest.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    
+    @Rule
+    public LoggerRule logs = new LoggerRule().record(Executor.class.getName(), Level.SEVERE).capture(100); //just in case future versions add more logs
+
+    public static final String ERROR_MESSAGE = "My unexpected exception";
+    
+    // When using SubTaskContributor (FailingSubTaskContributor) the build never ends
+    @Test
+    @Issue("JENKINS-59793")
+    public void buildDoesntFinishWhenSubTaskFails() throws Exception {
```
Comment (varyvol) [inline test/src/test/java/hudson/model/queue/BuildKeepsRunningWhenFaultySubTasksTest.java]:
```suggestion
    public void buildFinishesWhenSubTaskFails() throws Exception {
```
I think the name is appropriate before the fix but after the fix it should say what's actually asserting.

File: test/src/test/java/hudson/model/queue/BuildKeepsRunningWhenFaultySubTasksTest.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        FreeStyleProject p = j.createProject(FreeStyleProject.class);
+        QueueTaskFuture<FreeStyleBuild> future = p.scheduleBuild2(0);
+        assertThat("Build should be actually scheduled by Jenkins", future, notNullValue());
+
+        // We get stalled waiting the finalization of the job
+        FreeStyleBuild build = future.get(5, TimeUnit.SECONDS);
```
Comment (varyvol) [inline test/src/test/java/hudson/model/queue/BuildKeepsRunningWhenFaultySubTasksTest.java]: Should you also check the build result is a failure?

**Review by varyvol — COMMENTED** [review]
For all reviews, note this issue has been happening also in http://ci.jenkins.io

@MRamonLeon my doubt here is: why has the issue has started appearing now? The metrics plugin does not seem to have important changes related to this.

File: test/src/test/java/hudson/model/queue/BuildKeepsRunningWhenFaultySubTasksTest.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        assertTrue("The message is printed in the log", logs.getRecords().stream().anyMatch(r -> r.getThrown().getMessage().equals(ERROR_MESSAGE)));
```
[the remark this answers is not in the document]
Comment (MRamonLeon) [reply test/src/test/java/hudson/model/queue/BuildKeepsRunningWhenFaultySubTasksTest.java]: Yes, not needed if the code passes the `get`. It's a reminiscence from a lot of tests I've been doing. I can remove it.

File: test/src/test/java/hudson/model/queue/BuildKeepsRunningWhenFaultySubTasksTest.java (hunk 1 above)
Comment (MRamonLeon) [reply test/src/test/java/hudson/model/queue/BuildKeepsRunningWhenFaultySubTasksTest.java]: Pushed with the other change

File: test/src/test/java/hudson/model/queue/BuildKeepsRunningWhenFaultySubTasksTest.java (hunk 2 above)
Comment (MRamonLeon) [reply test/src/test/java/hudson/model/queue/BuildKeepsRunningWhenFaultySubTasksTest.java]: No need to assert that, the purpose of the test is to guarantee we pass this point, so let's keep the minimum needed as agreed above.

**Review by varyvol — APPROVED** [review]

**Review by timja — APPROVED** [review]

File: core/src/main/java/hudson/model/queue/WorkUnitContext.java [hunk 4]
```diff
@@ -108,12 +108,17 @@ public WorkUnit getPrimaryWorkUnit() {
      * All the {@link Executor}s that jointly execute a {@link Task} call this method to synchronize on the start.
      */
     public void synchronizeStart() throws InterruptedException {
```
Comment (stephenc) [inline core/src/main/java/hudson/model/queue/WorkUnitContext.java]: I think we want to set the exception state of the future for exceptional exits.

Also we can determine if this is the main work unit up front as none of the ISE exceptions in my suggestion will ever be thrown by the only internal code path
```suggestion
    /**
     * All the {@link Executor}s that jointly execute a {@link Task} call this method to synchronize on the start.
     */
    @Restricted(NoExternalUse.class)
    public void synchronizeStart() throws InterruptedException {
        Executor e = Executor.currentExecutor();
        if (e == null) {
            throw new IllegalStateException("synchronizeStart shall only be called from an Executor");
        }
        WorkUnit wu = e.getCurrentWorkUnit();
        if (wu == null) {
            throw new IllegalStateException("synchronizeStart shall only be called after a WorkUnit has been assigned");
        }
        boolean mainWork = wu.isMainWork();
        try {
            startLatch.synchronize();
            // the main thread will send a notification
            if (mainWork) {
                future.start.set(e.getCurrentExecutable());
            }
        } catch (Error | RuntimeException | InterruptedException problem) {
            if (mainWork) {
                future.start.set(problem);
            }
            throw problem;
        } finally {
            if (mainWork && !future.start.isDone()) {
                future.start.set(e.getCurrentExecutable());
            }
        }
    }
```

File: core/src/main/java/hudson/model/queue/WorkUnitContext.java [hunk 5]
```diff
@@ -108,12 +108,17 @@ public WorkUnit getPrimaryWorkUnit() {
...
+            startLatch.synchronize();
+        } finally {
+            // the main thread will send a notification
+            Executor e = Executor.currentExecutor();
+            WorkUnit wu = e.getCurrentWorkUnit();
+            if (wu.isMainWork()) {
+                future.start.set(e.getCurrentExecutable());
+            }
         }
     }
```
Comment (stephenc) [inline core/src/main/java/hudson/model/queue/WorkUnitContext.java]: We should block out any direct use of this method, `DoNotUse.class` might be appropriate

```suggestion

    @Restricted(NoExternalUse.class)
```

File: core/src/main/java/hudson/model/queue/WorkUnitContext.java [hunk 6]
```diff
@@ -130,17 +135,21 @@ public void synchronizeEnd(Queue.Executable executable, Throwable problems, long
      *      the member threads will report {@link InterruptedException}.
      */
     public void synchronizeEnd(Executor e, Queue.Executable executable, Throwable problems, long duration) throws InterruptedException {
```
Comment (stephenc) [inline core/src/main/java/hudson/model/queue/WorkUnitContext.java]: Again, would be better to set the future for an exception where an exception has occurred
```suggestion
    @Restricted(NoExternalUse.class)
    public void synchronizeEnd(Executor e, Queue.Executable executable, Throwable problems, long duration) throws InterruptedException {
        WorkUnit wu = e.getCurrentWorkUnit();
        if (wu == null) {
            throw new IllegalStateException("synchronizeStart shall only be called after a WorkUnit has been assigned");
        }
        boolean mainWork = wu.isMainWork();
        try {
            endLatch.synchronize();

            // the main thread will send a notification
            if (mainWork) {
                if (problems == null) {
                    future.set(executable);
                    e.getOwner().taskCompleted(e, task, duration);
                } else {
                    future.set(problems);
                    e.getOwner().taskCompletedWithProblems(e, task, duration, problems);
                }
            }
        } catch (Error | RuntimeException | InterruptedException problem) {
            if (mainWork) {
                future.set(problem);
            }
            throw problem;
        } finally {
            if (mainWork && !future.isDone()) {
                future.set(executable);
            }
        }
    }
```

File: core/src/main/java/hudson/model/queue/WorkUnitContext.java [hunk 7]
```diff
@@ -130,17 +135,21 @@ public void synchronizeEnd(Queue.Executable executable, Throwable problems, long
...
+                if (problems == null) {
+                    future.set(executable);
+                    e.getOwner().taskCompleted(e, task, duration);
+                } else {
+                    future.set(problems);
+                    e.getOwner().taskCompletedWithProblems(e, task, duration, problems);
+                }
             }
         }
     }
```
Comment (stephenc) [inline core/src/main/java/hudson/model/queue/WorkUnitContext.java]: I think the real issue is the `abort(Throwable)` method as that could exit also without completing the task, I think something like:

```suggestion
    }


    /**
     * When one of the work unit is aborted, call this method to abort all the other work units.
     */
    public synchronized void abort(Throwable cause) {
        if (cause==null)        throw new IllegalArgumentException();
        if (aborted!=null)      return; // already aborted    
        try {
            aborted = cause;
            startLatch.abort(cause);
            endLatch.abort(cause);

            Thread c = Thread.currentThread();
            for (WorkUnit wu : workUnits) {
                Executor e = wu.getExecutor();
                if (e != null && e != c)
                    e.interrupt();
            }
        } finally {
            if (!future.start.isDone()) {
                future.start.set(cause);
            }
            if (!future.isDone()) {
                future.set(cause);
            }
        }
    }
```

**Review by stephenc — APPROVED** [review]
Some suggestions for even more robustness and perhaps covering another edge case.

Your change makes things better (hence approving) so if you would rather merge your change first to get an improvement out and then examine my proposal in a subsequent PR that's totally fine

File: core/src/main/java/hudson/model/queue/WorkUnitContext.java (hunk 6 above)
Comment (oleg-nenashev) [reply core/src/main/java/hudson/model/queue/WorkUnitContext.java]: Let's throw a checked exception then so that we ensure it is proper handling in calling methods for the future

Comment (MRamonLeon) [reply core/src/main/java/hudson/model/queue/WorkUnitContext.java]: Which one are you suggesting? just `new InterruptedException(problem)`? It's not an interrupted exception actually.

**Comment by MRamonLeon** [comment]
Added @stephenc's suggetions

**Review by stephenc — APPROVED** [review]
LGTM

File: core/src/main/java/hudson/model/queue/WorkUnitContext.java [hunk 8]
```diff
@@ -107,13 +109,32 @@ public WorkUnit getPrimaryWorkUnit() {
...
+                future.start.set(e.getCurrentExecutable());
+            }
+        } catch (Error | RuntimeException | InterruptedException problem) {
+            if (mainWork) {
+                future.start.set(problem);
+            }
+            throw problem;
+        } finally {
+            if (mainWork && !future.start.isDone()) {
+                future.start.set(e.getCurrentExecutable());
```
Comment (varyvol) [inline core/src/main/java/hudson/model/queue/WorkUnitContext.java]: Won't this one override the one set if an exception is raised?

Comment (MRamonLeon) [reply core/src/main/java/hudson/model/queue/WorkUnitContext.java]: Yes, it overrides the same value. Anyway, I've changed the method code because we were at the same point, the new exceptions thrown when the _executable_ or the _current work uni_t is not set were keeping the `future` unset, so it may trigger the problem we were fixing after all.

**Comment by MRamonLeon** [comment]
I had to roll back to the original fix because the test was failing. Jenkins was shutting down after an unexpected exception in the test. I couldn't figure out why it was happening, so better we let this PR get merged and we improve it in a followup PR if desired.

**Review by rsandell — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
I plan to merge it tomorrow if no negative feedback

**Comment by batmat** [comment]
OK, moving on given no negative feedvback since yesterday.

Thank you eveyone!
